### ISSUE TITLE
Add custom YAML loading to check for duplicate keys

### ISSUE THREAD

**Issue by djhoese** [pytroll/satpy#1936]
One option to check for duplicate keys in the YAML files loaded/used by Satpy would be to add a custom constructor/loader as described in this gist:

https://gist.github.com/pypt/94d747fe5180851196eb

This wouldn't help the pre-commit in this PR, but at least the pre-commit is checking syntax.

_Originally posted by @djhoese in https://github.com/pytroll/satpy/issues/1935#issuecomment-992552084_

**Comment by djhoese** [comment]
That gist has comments with different options that should be considered as they may be better long term and more flexible. To clarify the above for any new contributors: Satpy uses YAML files as configuration for most of what it does. Inside these files we use special constructors to load python code directly. Because of this we have to use pyyamls "unsafe" loading which means that the python code is actually loaded. However, unsafe loading limits what type of checking/validation can be done automatically by some of the tools we use (like pre-commit mentioned in #1935 above).

Fixing this issue would do some checks when we load the YAML when using satpy, but not in a pre-commit check. If it is possible to do it with pre-commit that would be great. Another option might be to switch satpy YAML files to use string descriptions for the python objects instead of the fancy YAML constructors.

**Comment by mraspaud** [comment]
@djhoese I don't know if this is related, but I was thinking recently that providing files to read to satpy without specifying the reader to use takes nowadays quite some time. I believe this is mostly due to satpy instantiating all these python objects provided in the yaml file, while what it should really be doing first is to parse the filenames to see if they match the file patterns in the yaml files first. So maybe switching off the unsafe loading in a first step would be quite good for this too, and could help for yaml checking on top of that.

**Comment by djhoese** [comment]
Yeah that's a good point. I personally never use the "find a reader" functionality so I haven't experienced this too much. Unless I'm missing something it wouldn't be as simple as "switching off" the unsafe loading. We'd have to replace the python constructors in the YAML and then dynamically import the python code when it is needed. That or we do something like what we do in this yaml test here:

https://github.com/pytroll/satpy/blob/267e779cea71cb5286d5edc1b4da7478eb62fc62/satpy/tests/test_readers.py#L628-L635

where we pass `IgnoreLoader` as the loader that pyyaml should use.

**Comment by mraspaud** [comment]
Ideally we could have a LazyLoader, that doesn't instantiate anything until the value is used. But maybe I'm drifting a bit here
